document the new `analysis-phonenumber` plugin (#8469)
opensearch-project/documentation-website: _analyzers/supported-analyzers/phone-analyzers.md

## Example

The following request creates an index containing one field, which ingests phone numbers for Switzerland (region code `CH`):

```json
PUT /example-phone
{
  "settings": {
    "analysis": {
      "analyzer": {
        "phone-ch": {
          "type": "phone",
          "phone-region": "CH"
        },
        "phone-search-ch": {
          "type": "phone-search",
          "phone-region": "CH"
        }
      }
    }
  },
  "mappings": {
    "properties": {
      "phoneNumber": {
        "type": "text",
        "analyzer": "phone-ch",
        "search_analyzer": "phone-search-ch"
      }
    }
  }
}
```
{% include copy-curl.html %}

Analysing a (fictional) Swiss phone number with an international calling prefix will work the same with either the Swiss-specific phone region or without:
```json
GET /example-phone/_analyze
{
  "analyzer" : "phone-ch",
  "text" : "+41 60 555 12 34"
}
```
{% include copy-curl.html %}

and

```json
GET /example-phone/_analyze
{
  "analyzer" : "phone",
  "text" : "+41 60 555 12 34"
}
```
{% include copy-curl.html %}

will produce the same result:
```json

> ["+41 60 555 12 34", "6055512", "41605551", "416055512", "6055", "41605551234", ...]

Author: i used `...` here as i don't think there's much need in adding a ton of tokens to the example?

here's the complete output for the request above:
```json
{
    "tokens": [
        {
            "token": "+41 60 555 12 34",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 0
        },
        {
            "token": "6055512",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 1
        },
        {
            "token": "41605551",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 2
        },
        {
            "token": "416055512",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 3
        },
        {
            "token": "6055",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 4
        },
        {
            "token": "41605551234",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 5
        },
        {
            "token": "60555123",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 6
        },
        {
            "token": "41605",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 7
        },
        {
            "token": "4160",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 8
        },
        {
            "token": "4160555",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 9
        },
        {
            "token": "416055",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 10
        },
        {
            "token": "41 60 555 12 34",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 11
        },
        {
            "token": "60555",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 12
        },
        {
            "token": "605551",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 13
        },
        {
            "token": "6",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 14
        },
        {
            "token": "4160555123",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 15
        },
        {
            "token": "605551234",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 16
        },
        {
            "token": "416",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 17
        },
        {
            "token": "60",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 18
        },
        {
            "token": "605",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 19
        },
        {
            "token": "41",
            "start_offset": 0,
            "end_offset": 0,
            "type": "word",
            "position": 20
        }
    ]
}
```

Reviewer: Yes, that's perfect. Thank you for adding the examples! I rewrote a bit, and I think we're ready to move this to editorial review, after which we can merge the PR.

---
layout: default
title: Phone number
parent: Analyzers
nav_order: 140
---

# Phone number analyzers

> The `analysis-phonenumber` plugin provides analyzers and tokenizers for parsing phone numbers. A dedicated analyzer is required because parsing phone numbers is a non-trivial task (even though it might seem trivial at first glance). For common pitfalls in parsing phone numbers, see [Falsehoods programmers believe about phone numbers](https://github.com/google/libphonenumber/blob/master/FALSEHOODS.md).

Reviewer: ```suggestion
The `analysis-phonenumber` plugin provides analyzers and tokenizers for parsing phone numbers. A dedicated analyzer is required because parsing phone numbers is a non-trivial task (even though it might seem trivial at first glance). For common misconceptions regarding phone number parsing, see [Falsehoods programmers believe about phone numbers](https://github.com/google/libphonenumber/blob/master/FALSEHOODS.md).
```

> The `analysis-phonenumber` plugin provides analyzers and tokenizers for parsing phone numbers. A dedicated analyzer is required because parsing phone numbers is a non-trivial task (even though it might seem trivial at first glance). For common pitfalls in parsing phone numbers, see [Falsehoods programmers believe about phone numbers](https://github.com/google/libphonenumber/blob/master/FALSEHOODS.md).

Reviewer: Reworded.

OpenSearch supports the following phone number analyzers:

* [`phone`](#the-phone-analyzer): An [index analyzer]({{site.url}}{{site.baseurl}}/analyzers/index-analyzers/) to use at indexing time.
* [`phone-search`](#the-phone-search-analyzer): A [search analyzer]({{site.url}}{{site.baseurl}}/analyzers/search-analyzers/) to use at search time.

Internally, the plugin uses the [`libphonenumber`](https://github.com/google/libphonenumber) library and follows its parsing rules.

The phone number analyzers are not meant to find phone numbers in larger texts. Instead, you should use them on fields that only contain phone numbers.
{: .note}

## Installing the plugin

Before you can use the phone number analyzers, you must install the `analysis-phonenumber` plugin by running the following command:

```sh
./bin/opensearch-plugin install analysis-phonenumber
```

## Specifying a default region

> You can optionally specify a default region for parsing phone numbers by providing the `phone-region` parameter within the analyzer. Valid phone regions are represented by ISO 3166 country codes. For more information, see [List of ISO 3166 country codes](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes).

Author: i'm not a native speaker, but i'd have expected "[..] the list [..]" here? (it probably was me who wrote it like this in the first place? 🫣)
```suggestion
You can optionally specify a default region for parsing phone numbers by providing the `phone-region` parameter within the analyzer. Valid phone regions are represented by ISO 3166 country codes. For more information, see the [list of ISO 3166 country codes](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes).
```

Reviewer 2: @rursprung Either way is correct 😄

When tokenizing phone numbers containing the international calling prefix `+`, the default region is irrelevant. However, for phone numbers which either use a national prefix for international numbers (for example,  `001` instead of `+1` to dial Northern America from most European countries), the region needs to be provided. You can also properly index local phone numbers with no international prefix if you specify the region.

## Example

> The following request creates an index containing one field, which ingests phone numbers for Switzerland (region code `CH`):

Reviewer: ```suggestion
The following request creates an index containing one field that ingests phone numbers for Switzerland (region code `CH`):
```

```json
PUT /example-phone
{
  "settings": {
    "analysis": {
      "analyzer": {
        "phone-ch": {
          "type": "phone",
          "phone-region": "CH"
        },
        "phone-search-ch": {
          "type": "phone-search",
          "phone-region": "CH"
        }
      }
    }
  },
  "mappings": {
    "properties": {

>       "phoneNumber": {

Author: btw: do examples usually use `camelCase`, `snake_case` or `kebab-case`? this one currently uses `kebab-case` for the analyzer names (the actual `phone-search` analyzer does so too) and index name but `camelCase` for the field here (sorry, don't know how i ended up with this mix in this example)

Reviewer: In general, OpenSearch uses `snake_case` in requests.

Reviewer: Ideally, it should all be snake case, so yes, `phone_number` is preferred. That said, even our own alerting plugin uses camel case. I think it would be nice to change to snake case 😄

>       "phoneNumber": {

Reviewer: ```suggestion
      "phone_number": {
```

      }
    }
  }
}
```
{% include copy-curl.html %}

## The phone analyzer

The `phone` analyzer generates n-grams based on the given phone number. A (fictional) Swiss phone number containing an international calling prefix can be parsed with or without the Swiss-specific phone region. Thus, the following two requests will produce the same result:

```json
GET /example-phone/_analyze
{
  "analyzer" : "phone-ch",
  "text" : "+41 60 555 12 34"
}
```
{% include copy-curl.html %}

```json
GET /example-phone/_analyze
{
  "analyzer" : "phone",
  "text" : "+41 60 555 12 34"
}
```
{% include copy-curl.html %}

The response contains the generated n-grams:

```json
["+41 60 555 12 34", "6055512", "41605551", "416055512", "6055", "41605551234", ...]
```

However, if you specify the phone number without the international calling prefix `+` (either by using `0041` or omitting
the international calling prefix altogether), then only the analyzer configured with the correct phone region can parse the number:

```json
GET /example-phone/_analyze
{
  "analyzer" : "phone-ch",
  "text" : "060 555 12 34"
}
```
{% include copy-curl.html %}

## The phone-search analyzer

> In contrast, the `phone-search` analyzer does not create n-grams and only issues some basic tokens. Thus, the following request:

Reviewer: ```suggestion
In contrast, the `phone-search` analyzer does not create n-grams and only issues some basic tokens. For example, send the following request and specify the `phone-search` analyzer:
```

```json
GET /example-phone/_analyze
{
  "analyzer" : "phone-search",
  "text" : "+41 60 555 12 34"
}
```
{% include copy-curl.html %}

> is parsed into the following tokens:

Reviewer 2: "Is"?

> is parsed into the following tokens:

Reviewer: ```suggestion
Is parsed into the following tokens:
```

> Is parsed into the following tokens:

Author: i had `is` lowercase as it's the continuation of line 113. now it feels like a new sentence but isn't one (lacking a subject)

Reviewer 2: @rursprung I'm not a huge fan of the split-sentence structure, generally, and would prefer that we not use it because it results in this type of issue (case in point: I'd prefer that new lines/sentences/phrases not begin with lowercase letters.

> Is parsed into the following tokens:

Reviewer: ```suggestion
The response contains the following tokens:
```
